### ISSUE TITLE
error in finalize w/o `theme_cmap`

### ISSUE THREAD

**Issue by matthewstern** [CMAP-REPOS/cmapplot#83]
Finalize fails for me due to a lack of unit object sometimes--it looks like this happens specifically when passing a non-theme-cmap() plot to finalize(). While this is an unlikely occurrence, it's not preferable and it would be nice to allow the user to do this if they wanted. It would be good to explore this and root out the problem if possible.

My initial guesses is this has something to do with the various theme modifications that work within finalize, and/or the legend extraction process.

**Comment by nmpeterson** [comment]
I ran a traceback and the error is coming from the "preformat plot" section of finalize_plot.R.

```r
  # preformat plot
  plot <- plot + ggplot2::theme(
    # remove any in-plot titles
    plot.title = element_blank(),
    plot.caption = element_blank(),
    # add in legend_bump
    legend.margin = margin(
       l = grid::convertUnit(plot$theme$legend.margin[[4]], "bigpts", valueOnly = TRUE) +
         consts$legend_bump,
       unit = "bigpts"),
    # apply any extra `ggplot2::theme()` args
    ...
  )
```

Specifically the `grid::convertUnit(plot$theme$legend.margin[[4]], "bigpts", valueOnly = TRUE)` part.

**Comment by matthewstern** [comment]
Good catch. It looks like a saved `plot` that had no theme applied does not have `$theme`, so definitely doesn't have `$theme$legend.margin[[4]]`. 

Applying `+ theme_grey()` to the plot fixes this problem, but creates another one in the code:
```
  margin_legend_i <- ifelse(
    # if not overridden in finalize, use value from ggplot
    is_null(overrides$margin_legend_i),
    plot$theme$legend.spacing.y,
    # otherwise, use override value
    overrides$margin_legend_i
  )
```
The issue here is that in the default theme, `$theme$legend.spacing.y` is NULL as `$theme$legend.spacing` is set instead.

To the first problem, I can think of a few possible fixes:
- we could build in an `ifelse` that extracts this value from ggplot's default theme (`ggplot_globals$theme_current` I think)
- we could apply a complete theme to `plot` within finalize (either the ggplot default or `theme_cmap()`)
- we could eliminate the `legend_bump` arg altogether as legend placement *should* be a non-issue at this point

To the second problem, I think we probably replace the if true argument (currently `plot$theme$legend.spacing.y`) with a `case_when`:
1. If `plot$theme$legend.spacing.y` exists, use it
2. If `plot$theme$legend.spacing` exists, use it
3. If `ggplot_globals$theme_current$legend.spacing.y` exists, use it
4. Otherwise, use `ggplot_globals$theme_current$legend.spacing`

What do you think @nmpeterson?

**Comment by nmpeterson** [comment]
Regarding the latter, I think the `case_when()` solution sounds good.

Wondering about @dlcomeaux's thoughts on eliminating `legend_bump` for the former?

**Comment by dlcomeaux** [comment]
I think we can go ahead and get rid of `legend_bump` - if people really want to modify it, they  should still be able to do so via the standard ggplot theme modifications. @nmpeterson 

The `case_when` makes sense to me as well.

**Comment by matthewstern** [comment]
You know, before the `# preformat plot` section as currently implemented, we could do:

```
if( is.na(plot$theme ){
   plot <- plot + ggplot_globals$theme_current
}
```

This would solve the first problem AND get rid of nums 3 and 4 from the case_when list.

**Comment by dlcomeaux** [comment]
@matthewstern that seems like it would work, and is worth it to simplify the cases. However, I think it might actually make sense to remove `legend_bump` even so - that argument was created to solve a problem that no longer exists, and I think to the degree we can simplify the functions we export, we should.

**Comment by matthewstern** [comment]
I agree! If someone wants to implement this, that would be swell. Otherwise, I can take a crack over the next few days. I'm ambivalent on explicitly adding the theme at this point. 

Note that the `case_when` is unnecessarily verbose considering the pattern is always "use X if it isn't NULL". This appears to work instead:
```
data.table::first(purrr::compact(list(
  plot$theme$legend.spacing.y,
  plot$theme$legend.spacing,
  ggplot2:::ggplot_global$theme_current$legend.spacing.y,
  ggplot2:::ggplot_global$theme_current$legend.spacing
)))
```

**Comment by dlcomeaux** [comment]
Both issues are now addressed in #67 so we are set once that is merged.
